airflow/providers/common/sql/hooks/sql.py

```diff
@@ -232,22 +270,19 @@ def run(self, sql, autocommit=False, parameters=None, handler=None):
                 results = []
                 for sql_statement in sql:
                     self._run_command(cur, sql_statement, parameters)
+
                     if handler is not None:
                         result = handler(cur)
                         results.append(result)
 
-            # If autocommit was set to False for db that supports autocommit,
-            # or if db does not supports autocommit, we do a manual commit.
+            # If autocommit was set to False or db does not support autocommit, we do a manual commit.
             if not self.get_autocommit(conn):
                 conn.commit()
 
         if handler is None:
             return None
-
-        if scalar:
-            return results[0]
-
-        return results
+        else:
```

Comment:
I think this is backwards-incompatike. Currently when run gets a single SQL string, it return s single result, if it gets array of strings it returns array of results. In order to keep backwards compatibility we should retain this behaviour.

I think the reasonable approach will be to check if even after split we have one or multiple queries. If 1 -> return resullts[0] else results.

Reply:
hmm. I see that there are different behaviours in different operators. I think for now maybe we need another parameter (scalar_single_result: bool) ? this way we could keep full backwards compatibility in all the providers I think ?

Reply:
I think in the very first iteration it was `return results[-1]` which good for a single statement and list of them.
In one of our previous PRs, we got the idea to return all statement's results.
https://github.com/apache/airflow/issues/23623#issuecomment-1122758517
In this case, we will have the next returning type `Optional[List[HandlerResult]]`. 
Before that, it was `Optional[HandlerResult]`. We could add parameter `return_last=True` and then we will have `Optional[Union[HandlerResult, List[HandlerResult]]]`. For example:
if no handler then `-> None` else:
if one statement and `return_last=True` then `-> HandlerResult`
if `split_statements=True` and `return_last=True` then `-> HandlerResult` 
if `split_statements=True` and `return_last=False` then `-> List[HandlerResult]` 

also for one statement we could return a list with single value
if one statement and `return_last=False` then `-> List[HandlerResult]`

In code, I assume it'll be
```
if handler is None:
    return None
elif return_last:
    return results[-1]
else:
    return results
```

Reply:
`return_last` flag seems reasonable.

I do like to point to my note https://github.com/apache/airflow/issues/23623#issuecomment-1122973826 but that ship has sailed.